### PR TITLE
issue #646 Implement open file dialog for osx

### LINKED ISSUE #646: open file dialog not implemented for osx
## Expected Behavior
I expected open_file_dialog to work for osx.

## Current Behavior
```
File "<venv>/lib/python3.7/site-packages/toga/window.py", line 275, in open_file_dialog
    return self._impl.open_file_dialog(title, initial_directory, file_types, multiselect)
AttributeError: 'MainWindow' object has no attribute 'open_file_dialog'
```

## Steps to reproduce
`app.main_window.open_file_dialog(...`

## Your Environment
* Python Version
3.7.2

* Operating System and Version (select from the following and list the specific version number; if your OS is not listed, list that as well)

  - [x] macOS - version: 10.14.3
  - [ ] Linux - distro:   - version:
  - [ ] Windows - version:
  - [ ] Other - name:  - version:

* Toga Target (the type of app you are trying to generate)
 
  - [ ] android
  - [x] cocoa
  - [ ] django 
  - [ ] gtk
  - [ ] iOS
  - [ ] tvOS
  - [ ] watchOS
  - [ ] winforms 
  - [ ] win32
  - [ ] Other (please specify)

### PR DESCRIPTION
**Pull request by nehararora**
Added an open_file_dialog implementation the cocoa implementation. I've added a button to the dialog examples app as well.

Fixes #646 

## PR Checklist:

- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

File: src/cocoa/toga_cocoa/dialogs.py [hunk 1]
```diff
@@ -99,3 +99,37 @@ def save_file(window, title, suggested_filename, file_types=None):
     if result == NSFileHandlingPanelOKButton:
```
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/dialogs.py]: At column 18: (F405) 'NSFileHandlingPanelOKButton' may be undefined, or defined from star imports: .libs

File: src/cocoa/toga_cocoa/dialogs.py [hunk 2]
```diff
@@ -99,3 +99,37 @@ def save_file(window, title, suggested_filename, file_types=None):
...
+        window: The window this dialog belongs to.
+        title: Title of the modal.
+        file_types: Ignored for now.
+        multiselect: Flag to allow multiple file selection.
+    Returns: The file path on success, None otherwise
+
+    """
+
+    # Initialize and configure the panel.
+    panel = NSOpenPanel.alloc().init()
```
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/dialogs.py]: At column 13: (F405) 'NSOpenPanel' may be undefined, or defined from star imports: .libs

File: src/cocoa/toga_cocoa/dialogs.py [hunk 3]
```diff
@@ -99,3 +99,37 @@ def save_file(window, title, suggested_filename, file_types=None):
...
+    panel.title = title
+    panel.allowedFileTypes = file_types
+    panel.allowsMultipleSelection = multiselect
+    panel.canChooseDirectories = False
+    panel.canCreateDirectories = True
+    panel.canChooseFiles = True
+
+    # Show modal and return file path on success.
+    result = panel.runModal()
+    if NSFileHandlingPanelOKButton == result:
```
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/dialogs.py]: At column 8: (F405) 'NSFileHandlingPanelOKButton' may be undefined, or defined from star imports: .libs

**Comment by nehararora** [comment]
@danyeaw, @freakboy3742  - wasn't sure about allowing the creation of folders (osx shows a New Folder button if the flag's enabled) on open_file:
```
panel.canCreateDirectories = True 
```

For now it's enabled, I can take that out if that's not the behavior we want.

File: src/cocoa/toga_cocoa/dialogs.py (hunk 3 above)
Comment (yingw787) [inline src/cocoa/toga_cocoa/dialogs.py]: Just to leave my GitHub handle here

Comment (nehararora) [reply src/cocoa/toga_cocoa/dialogs.py]: @danyeaw, @freakboy3742  any conventions around this I should follow?
```
     if NSFileHandlingPanelOKButton == result: 
```
vs.
```
     if result ==  NSFileHandlingPanelOKButton: 
```

File: src/cocoa/toga_cocoa/dialogs.py [hunk 4]
```diff
@@ -99,3 +99,37 @@ def save_file(window, title, suggested_filename, file_types=None):
...
+
+    """
+
+    # Initialize and configure the panel.
+    panel = NSOpenPanel.alloc().init()
+    panel.title = title
+    panel.allowedFileTypes = file_types
+    panel.allowsMultipleSelection = multiselect
+    panel.canChooseDirectories = False
+    panel.canCreateDirectories = True
```
Comment (yingw787) [inline src/cocoa/toga_cocoa/dialogs.py]: @danyeaw @freakboy3742 offline discussion w/ @nehararora concluded that opening multiple file dialogs in macOS usually does allow for directory creation, is this correct assumption?

File: src/cocoa/toga_cocoa/dialogs.py [hunk 5]
```diff
@@ -99,3 +99,37 @@ def save_file(window, title, suggested_filename, file_types=None):
...
+
+    We restrict the panel invocation to only choose files. We also allow
+    creating directories but not selecting directories.
+
+    Args:
+        window: The window this dialog belongs to.
+        title: Title of the modal.
+        file_types: Ignored for now.
+        multiselect: Flag to allow multiple file selection.
+    Returns: The file path on success, None otherwise
```
Comment (danyeaw) [inline src/cocoa/toga_cocoa/dialogs.py]: The statement after `Returns:` should be on a new line with a period at the end of the sentence.

File: examples/dialogs/dialogs/app.py [hunk 6]
```diff
@@ -24,11 +24,22 @@ def action_open_file_dialog(self, widget):
         try:
             fname = self.main_window.open_file_dialog(
                 title="Open file with Toga",
+                multiselect=False
             )
             self.label.text = "File to open:" + fname
         except ValueError:
             self.label.text = "Open file dialog was canceled"
 
+    def action_open_file_dialog_multi(self, widget):
```
Comment (yingw787) [inline examples/dialogs/dialogs/app.py]: @nehararora did some manual testing and fname is None if the user exits the dialog w/o selecting a file, resulting in a TypeError when concatenating a str with NoneType in line 28. Could we include a ternary to cast fname to empty string if open_file_dialog() returns None, or return empty string within the method itself?

File: src/cocoa/toga_cocoa/dialogs.py (hunk 1 above)
Comment (nehararora) [reply src/cocoa/toga_cocoa/dialogs.py]: After talking to @danyeaw, not applicable - follows the existing convention of the * import.

File: src/cocoa/toga_cocoa/dialogs.py (hunk 2 above)
Comment (nehararora) [reply src/cocoa/toga_cocoa/dialogs.py]: After talking to @danyeaw, not applicable - follows the existing convention of the * import.

File: src/cocoa/toga_cocoa/dialogs.py (hunk 3 above)
Comment (nehararora) [reply src/cocoa/toga_cocoa/dialogs.py]: After talking to @danyeaw, not applicable - follows the existing convention of the * import.

File: src/cocoa/toga_cocoa/dialogs.py (hunk 1 above)
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/dialogs.py]: At column 18: (F405) 'NSFileHandlingPanelOKButton' may be undefined, or defined from star imports: .libs

File: src/cocoa/toga_cocoa/dialogs.py [hunk 7]
```diff
@@ -99,3 +99,37 @@ def save_file(window, title, suggested_filename, file_types=None):
...
+        window: The window this dialog belongs to.
+        title: Title of the modal.
+        file_types: Ignored for now.
+        multiselect: Flag to allow multiple file selection.
+    Returns:
+        The file path on success, None otherwise.
+    """
+
+    # Initialize and configure the panel.
+    panel = NSOpenPanel.alloc().init()
```
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/dialogs.py]: At column 13: (F405) 'NSOpenPanel' may be undefined, or defined from star imports: .libs

File: src/cocoa/toga_cocoa/dialogs.py [hunk 8]
```diff
@@ -99,3 +99,37 @@ def save_file(window, title, suggested_filename, file_types=None):
...
+    panel.title = title
+    panel.allowedFileTypes = file_types
+    panel.allowsMultipleSelection = multiselect
+    panel.canChooseDirectories = False
+    panel.canCreateDirectories = True
+    panel.canChooseFiles = True
+
+    # Show modal and return file path on success.
+    result = panel.runModal()
+    if NSFileHandlingPanelOKButton == result:
```
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/dialogs.py]: At column 8: (F405) 'NSFileHandlingPanelOKButton' may be undefined, or defined from star imports: .libs

File: src/cocoa/toga_cocoa/dialogs.py (hunk 3 above)
Comment (nehararora) [reply src/cocoa/toga_cocoa/dialogs.py]: changed order to follow convention

File: src/cocoa/toga_cocoa/dialogs.py (hunk 1 above)
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/dialogs.py]: At column 18: (F405) 'NSFileHandlingPanelOKButton' may be undefined, or defined from star imports: .libs

File: src/cocoa/toga_cocoa/dialogs.py (hunk 7 above)
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/dialogs.py]: At column 13: (F405) 'NSOpenPanel' may be undefined, or defined from star imports: .libs

File: src/cocoa/toga_cocoa/dialogs.py [hunk 9]
```diff
@@ -99,3 +99,37 @@ def save_file(window, title, suggested_filename, file_types=None):
...
+    panel.title = title
+    panel.allowedFileTypes = file_types
+    panel.allowsMultipleSelection = multiselect
+    panel.canChooseDirectories = False
+    panel.canCreateDirectories = True
+    panel.canChooseFiles = True
+
+    # Show modal and return file path on success.
+    result = panel.runModal()
+    if result == NSFileHandlingPanelOKButton:
```
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/dialogs.py]: At column 18: (F405) 'NSFileHandlingPanelOKButton' may be undefined, or defined from star imports: .libs

File: src/cocoa/toga_cocoa/dialogs.py (hunk 1 above)
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/dialogs.py]: At column 18: (F405) 'NSFileHandlingPanelOKButton' may be undefined, or defined from star imports: .libs

File: src/cocoa/toga_cocoa/dialogs.py (hunk 7 above)
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/dialogs.py]: At column 13: (F405) 'NSOpenPanel' may be undefined, or defined from star imports: .libs

File: src/cocoa/toga_cocoa/dialogs.py (hunk 9 above)
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/dialogs.py]: At column 18: (F405) 'NSFileHandlingPanelOKButton' may be undefined, or defined from star imports: .libs

File: src/cocoa/toga_cocoa/dialogs.py (hunk 1 above)
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/dialogs.py]: At column 18: (F405) 'NSFileHandlingPanelOKButton' may be undefined, or defined from star imports: .libs

File: src/cocoa/toga_cocoa/dialogs.py (hunk 7 above)
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/dialogs.py]: At column 13: (F405) 'NSOpenPanel' may be undefined, or defined from star imports: .libs

File: src/cocoa/toga_cocoa/dialogs.py (hunk 9 above)
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/dialogs.py]: At column 18: (F405) 'NSFileHandlingPanelOKButton' may be undefined, or defined from star imports: .libs

File: examples/dialogs/dialogs/app.py (hunk 6 above)
Comment (nehararora) [reply examples/dialogs/dialogs/app.py]: @yingw787 good catch! fixed.

File: src/cocoa/toga_cocoa/dialogs.py (hunk 1 above)
Comment (nehararora) [reply src/cocoa/toga_cocoa/dialogs.py]: not applicable

File: src/cocoa/toga_cocoa/dialogs.py (hunk 8 above)
Comment (nehararora) [reply src/cocoa/toga_cocoa/dialogs.py]: not applicable

File: src/cocoa/toga_cocoa/dialogs.py (hunk 1 above)
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/dialogs.py]: At column 18: (F405) 'NSFileHandlingPanelOKButton' may be undefined, or defined from star imports: .libs

File: src/cocoa/toga_cocoa/dialogs.py (hunk 7 above)
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/dialogs.py]: At column 13: (F405) 'NSOpenPanel' may be undefined, or defined from star imports: .libs

File: src/cocoa/toga_cocoa/dialogs.py [hunk 10]
```diff
@@ -99,3 +99,37 @@ def save_file(window, title, suggested_filename, file_types=None):
...
+    panel.title = title
+    panel.allowedFileTypes = file_types
+    panel.allowsMultipleSelection = multiselect
+    panel.canChooseDirectories = False
+    panel.canCreateDirectories = False
+    panel.canChooseFiles = True
+
+    # Show modal and return file path on success.
+    result = panel.runModal()
+    if result == NSFileHandlingPanelOKButton:
```
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/dialogs.py]: At column 18: (F405) 'NSFileHandlingPanelOKButton' may be undefined, or defined from star imports: .libs

File: src/cocoa/toga_cocoa/dialogs.py (hunk 4 above)
Comment (nehararora) [reply src/cocoa/toga_cocoa/dialogs.py]: @freakboy3742 , @danyeaw , @yingw787 I've changed to remove directory creation for now. We can always add it back in if it becomes something that hordes of people demand. :)

**Comment by danyeaw** [comment]
Thanks @nehararora @yingw787! :heart:

**Pull request merged**